Refactor and improve MLL tests

vitek-karas opened the pull request:
* Remove duplicated code - I know that tests should be descriptive, but repeating 100 times that we want to capture output doesn't feel necessary.
* For some of them move more stuff into the shared test state - this improves perf as we avoid repeating the setup (copying files around) for each test case, we do it once for the entire class
I also changed some of the tests to "Theory" as it's easier to read that way.
  * `SDKLookup.cs` - moved most of the state in to the shared state to speed up the tests.

* Adding more cases into the "theory" tests
  * Most notably for the framework resolution I added variations on the TFM (which will be needed when we implement disable of MLL)
* Adding new tests mostly around "list runtimes" (various forms), "list sdks" (various forms) and errors (which also list runtimes or sdks)

* Ported all of the `MultiLevelLookupFramework` tests over to the `FrameworkResolution` and `DependencyResolutions` suites which have a more robust test infra and can run the same tests much faster. Along the way I added lot more variations on top of the existing tests:
  * `PerAssemblyVersionResolutionMultipleFrameworks.cs` - this is actually not an MLL test, but I moved it to the new infra to make it much faster
  * `MultipleHives.cs` - MLL framework resolution tests

For SDK resolution I kept the `MultiLevelSDKLookup.cs` just removed code duplication and added new variants.

For the core reviewers: I promise I didn't remove any single test case in spirit with these exceptions:

* We had tests which validated that framework resolution is not affected by frameworks in current directory and also by frameworks in the user's directory. I left some basic test for the current directory check, but I completely removed the user's directory variant as the product simply doesn't have any code around that anymore.

vitek-karas committed:
Refactor and improve MLL tests

* Remove duplicated code - I know that tests should be descriptive, but repeating 100 times that we want to capture output doesn't feel necessary.
* For some of them move more stuff into the shared test state - this improves perf as we avoid repeating the setup (copying files around) for each test case, we do it once for the entire class
I also changed some of the tests to "Theory" as it's easier to read that way.
  * `SDKLookup.cs` - moved most of the state in to the shared state to speed up the tests.

* Adding more cases into the "theory" tests
  * Most notably for the framework resolution I added variations on the TFM (which will be needed when we implement disable of MLL)
* Adding new tests mostly around "list runtimes" (various forms), "list sdks" (various forms) and errors (which also list runtimes or sdks)

* Ported all of the `MultiLevelLookupFramework` tests over to the `FrameworkResolution` and `DependencyResolutions` suites which have a more robust test infra and can run the same tests much faster. Along the way I added lot more variations on top of the existing tests:
  * `PerAssemblyVersionResolutionMultipleFrameworks.cs` - this is actually not an MLL test, but I moved it to the new infra to make it much faster
  * `MultipleHives.cs` - MLL framework resolution tests

For SDK resolution I kept the `MultiLevelSDKLookup.cs` just removed code duplication and added new variants.

For the core reviewers: I promise I didn't remove any single test case in spirit with these exceptions:

* We had tests which validated that framework resolution is not affected by frameworks in current directory and also by frameworks in the user's directory. I left some basic test for the current directory check, but I completely removed the user's directory variant as the product simply doesn't have any code around that anymore.

elinor-fung commented:
> completely removed the user's directory variant as the product simply doesn't have any code around that anymore

Could we do this for the SDK lookup tests too? I don't recall ever seeing anything around user directory for SDK resolution either.

vitek-karas committed:
Remove "user" directory from SDK resolution tests

The product simply doesn't have any knoweldge about user directory SDK lookup, so there's little value in testing for it (even though the test expects it to not have any impact).

vitek-karas commented:
Good point - I removed the "user directory" parts from SDK resolution tests as well.

elinor-fung commented on src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:
```diff
@@ -0,0 +1,200 @@
+
+        [Theory]
+        [InlineData("1.1.1")]  // Exact match - no roll forward
+        [InlineData("1.1.0")]  // Path roll forward
```
```suggestion
        [InlineData("1.1.0")]  // Patch roll forward
```

elinor-fung commented on src/installer/tests/HostActivation.Tests/FrameworkResolution/MultipleHives.cs:
```diff
@@ -20,56 +22,211 @@ public MultipleHives(SharedTestState sharedState)
+            RunTest(new TestSettings()
+                    .WithCommandLine("--list-runtimes"),
+                multiLevelLookup,
+                testApp: null)
```
nit: indentation

elinor-fung commented on src/installer/tests/HostActivation.Tests/DependencyResolution/PerAssemblyVersionResolutionMultipleFrameworks.cs:
```diff
@@ -0,0 +1,200 @@
+        [InlineData(TestAssemblyWithFileVersion, "1.0.0.0", null, null)] // App has assembly version, no framework has it - so app wins
+        [InlineData(TestAssemblyWithFileVersion, null, "1.0.0.0", MicrosoftNETCoreApp)]
+        [InlineData(TestAssemblyWithFileVersion, null, "4.0.0.0", null)] // App has higher version than either framework
+        [InlineData(TestAssemblyWithFileVersion, null, "3.2.2.2", MicrosoftNETCoreApp)] // Exactly equal - lower one wins
```
I feel like I've asked this before, but what does `lower` mean here?

elinor-fung replied:
Oh, this is the dependency hierarchy / depth thing, right? So app is highest level and NETCoreApp would always be lowest?

vitek-karas replied:
Yes - the app is the highest level, NETCoreApp is always the "root" lowest level. In the resolutio logic, if everything else is equal the lowest level typically wins.

vitek-karas committed:
Apply suggestions from code review

Co-authored-by: Elinor Fung <elfung@microsoft.com>

vitek-karas commented on src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs:
```diff
@@ -679,6 +467,103 @@ public void SdkMultilevelLookup_Must_Pick_The_Highest_Semantic_Version()
+                expectedList.Add(("6.2.0", _regSdkBaseDir));
+                expectedList.Add(("7.0.1", _regSdkBaseDir));
+            }
+            expectedList.Sort((a, b) => a.version.CompareTo(b.version));
```
Good point - fixed.

vitek-karas commented on src/installer/tests/HostActivation.Tests/MultilevelSDKLookup.cs:
```diff
@@ -679,6 +467,103 @@ public void SdkMultilevelLookup_Must_Pick_The_Highest_Semantic_Version()
+            var expectedList = AddSdkVersionsAndGetExpectedList(multiLevelLookup);
+            string expectedOutput = String.Join(String.Empty, expectedList.Select(t => $"{t.version} [{t.rootPath}]{Environment.NewLine}"));
+
+            RunTest("--list-sdks", multiLevelLookup)
```
Good idea - added.

vitek-karas committed:
PR Feedback

vitek-karas commented:
The test failures are:
https://github.com/dotnet/runtime/issues/66174
https://github.com/dotnet/runtime/issues/65954
